modules/renter/contractor/uptime.go

```diff
@@ -0,0 +1,94 @@
+package contractor
+
+import (
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+)
+
+// uptimeInterval specifies how frequently hosts are checked.
+var (
+	uptimeInterval = func() time.Duration {
+		switch build.Release {
+		case "dev":
+			return 1 * time.Minute
+		case "standard":
+			return 30 * time.Minute
+		case "testing":
+			return 100 * time.Millisecond
+		}
+		panic("undefined uptimeInterval")
+	}()
+
+	// uptimeWindow specifies the duration in which host uptime is checked.
+	uptimeWindow = func() time.Duration {
+		switch build.Release {
+		case "dev":
+			return 24 * time.Hour // 1 day
+		case "standard":
+			return 7 * 24 * time.Hour // 1 week
+		case "testing":
+			return 1 * time.Minute // 1 minute
+		}
+		panic("undefined uptimeWindow")
+	}()
+)
+
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) it has been scanned at least 3 times within the uptime window, and
+	// 2) the 3 most recent scans all failed
+	var lastWeek []modules.HostDBScan
+	for _, scan := range host.ScanHistory {
+		if time.Since(scan.Timestamp) < uptimeWindow {
+			lastWeek = append(lastWeek, scan)
+		}
+	}
+	// need at least 3 scans to make a fair judgment
+	if len(lastWeek) < 3 {
+		return false
+	}
+	// return true if all 3 most recent scans failed
+	// NOTE: lastWeek is ordered from oldest to newest
+	lastWeek = lastWeek[len(lastWeek)-3:]
+	return !lastWeek[0].Success && !lastWeek[1].Success && !lastWeek[2].Success
```

Comment:
*

2. There are a minimum of 3 data points total AND the 3 most recent data points all indicate that the host is offline.